diff -u Makefile.am.orig Makefile.am
--- fontconfig-2.12.1/Makefile.am.orig	2016-07-10 22:41:05.000000000 -0400
+++ fontconfig-2.12.1/Makefile.am	2016-12-24 13:06:05.000000000 -0500
@@ -65,9 +65,11 @@
 xml_DATA = fonts.dtd
 
 if CROSS_COMPILING
-  RUN_FC_CACHE_TEST=false
+	RUN_FC_CACHE_TEST=false
 else
-  RUN_FC_CACHE_TEST=test -z "$(DESTDIR)"
+	ifndef RUN_FC_CACHE_TEST
+		RUN_FC_CACHE_TEST=test -z "$(DESTDIR)"
+	endif
 endif
 
 # Creating ChangeLog from git log:
