diff -Naur fontconfig-2.11.94.orig/src/fccache.c fontconfig-2.11.94/src/fccache.c
--- fontconfig-2.11.94.orig/src/fccache.c	2015-05-27 23:53:27 +0000
+++ fontconfig-2.11.94/src/fccache.c	2016-06-15 12:00:00 +0000
@@ -34,6 +34,7 @@
 #  include <unistd.h>
 #  include <sys/mman.h>
 #endif
+#include <emmintrin.h>
 
 #ifndef O_BINARY
 #define O_BINARY 0
diff -Naur fontconfig-2.11.94.orig/src/fccfg.c fontconfig-2.11.94/src/fccfg.c
--- fontconfig-2.11.94.orig/src/fccfg.c	2015-04-09 00:19:17 +0000
+++ fontconfig-2.11.94/src/fccfg.c	2014-06-15 12:00:00 +0000
@@ -27,6 +27,7 @@
 #include "fcint.h"
 #include <dirent.h>
 #include <sys/types.h>
+#include <emmintrin.h>
 #include "../fc-blanks/fcblanks.h"
 
 #if defined (_WIN32) && !defined (R_OK)
diff -Naur fontconfig-2.11.94.orig/src/fcxml.c fontconfig-2.11.94/src/fcxml.c 
--- fontconfig-2.11.94.orig/src/fcxml.c	2015-05-28 05:53:27.000000000 +0000
+++ fontconfig-2.11.94/src/fcxml.c	2016-06-15 12:00:00 +0000
@@ -26,6 +26,7 @@
 #include <fcntl.h>
 #include <stdarg.h>
 #include <dirent.h>
+#include <emmintrin.h>
 
 #ifdef ENABLE_LIBXML2
 
diff -Naur fontconfig-2.11.0/src/fcdefault.c fontconfig-2.11.0.new/src/fcdefault.c
--- fontconfig-2.11.0/src/fcdefault.c	2013-10-11 03:10:18 +0000
+++ fontconfig-2.11.0.new/src/fcdefault.c	2014-02-15 18:44:53 +0000
@@ -25,6 +25,7 @@
 #include "fcint.h"
 #include <limits.h>
 #include <string.h>
+#include <emmintrin.h>
 
 /* MT-safe */
 
diff -Naur fontconfig-2.11.0/src/fcobjs.c fontconfig-2.11.0.new/src/fcobjs.c
--- fontconfig-2.11.0/src/fcobjs.c	2013-10-11 03:10:18 +0000
+++ fontconfig-2.11.0.new/src/fcobjs.c	2014-02-15 18:45:17 +0000
@@ -33,6 +33,7 @@
 #include "fcobjshash.h"
 
 #include <string.h>
+#include <emmintrin.h>
 
 /* The 1000 is to leave some room for future added internal objects, such
  * that caches from newer fontconfig can still be used with older fontconfig

