From 16d4559b01cb04b55b9af987538d04012228e551 Mon Sep 17 00:00:00 2001
From: Damiano Galassi <damiog@gmail.com>
Date: Sat, 20 Apr 2024 13:43:46 +0200
Subject: [PATCH] Expose parse_itu_t35_dovi_metadata_obu to C api.

---
 dolby_vision/src/capi.rs | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/dolby_vision/src/capi.rs b/dolby_vision/src/capi.rs
index 5b425e3..878cda8 100644
--- a/dolby_vision/src/capi.rs
+++ b/dolby_vision/src/capi.rs
@@ -27,6 +27,21 @@ pub unsafe extern "C" fn dovi_parse_rpu(buf: *const u8, len: size_t) -> *mut Rpu
     Box::into_raw(Box::new(RpuOpaque::from(res)))
 }
 
+/// # Safety
+/// The pointer to the data must be valid.
+///
+/// Parse a Dolby Vision from a AV1 ITU-T T.35 metadata OBU byte buffer.
+/// Adds an error if the parsing fails.
+#[no_mangle]
+pub unsafe extern "C" fn dovi_parse_itu_t35_dovi_metadata_obu(buf: *const u8, len: size_t) -> *mut RpuOpaque {
+    assert!(!buf.is_null());
+
+    let data = slice::from_raw_parts(buf, len);
+    let res = DoviRpu::parse_itu_t35_dovi_metadata_obu(data);
+
+    Box::into_raw(Box::new(RpuOpaque::from(res)))
+}
+
 /// # Safety
 /// The pointer to the data must be valid.
 ///
-- 
2.39.3 (Apple Git-146)

