diff -Naurw xvidcore-20090311.orig/build/generic/bootstrap.sh xvidcore-20090311/build/generic/bootstrap.sh
--- xvidcore-20090311.orig/build/generic/bootstrap.sh	2005-05-23 05:29:43.000000000 -0400
+++ xvidcore-20090311/build/generic/bootstrap.sh	2009-04-15 13:13:54.000000000 -0400
@@ -75,7 +75,7 @@
 $AUTOMAKE -c -a 1>/dev/null 2>&1
 
 echo "Copying files provided by libtool"
-$LIBTOOLIZE -f -c 1>/dev/null 2>&1
+$LIBTOOLIZE -f -c -i 1>/dev/null 2>&1
 
 echo "Removing files that are not needed"
 rm -rf autom4* 1>/dev/null 2>&1 
