diff -Naur xvidcore-20090311.orig/build/generic/configure.in xvidcore-20090311/build/generic/configure.in
--- xvidcore-20090311.orig/build/generic/configure.in	2008-12-05 05:18:52.000000000 -0500
+++ xvidcore-20090311/build/generic/configure.in	2009-04-14 11:30:04.000000000 -0400
@@ -427,8 +427,13 @@
 	       MARK_FUNCS=""
                ;;
            *darwin*)
-               AC_MSG_RESULT([macho])
-               NASM_FORMAT="macho"
+               if test "$ARCHITECTURE" = "X86_64" ; then
+                   AC_MSG_RESULT([macho64])
+                   NASM_FORMAT="macho64"
+               else
+                   AC_MSG_RESULT([macho])
+                   NASM_FORMAT="macho"
+               fi
                PREFIX="-DPREFIX"
                MARK_FUNCS=""
                ;;
diff -Naur xvidcore-20090311.orig/src/nasm.inc xvidcore-20090311/src/nasm.inc
--- xvidcore-20090311.orig/src/nasm.inc	2008-12-05 05:15:02.000000000 -0500
+++ xvidcore-20090311/src/nasm.inc	2009-04-14 12:13:37.000000000 -0400
@@ -177,6 +177,8 @@
 %macro TEXT 0
 %ifidn __OUTPUT_FORMAT__,macho
       SECTION .text align=SECTION_ALIGN
+%elifidn __OUTPUT_FORMAT__,macho64
+      SECTION .text align=SECTION_ALIGN
 %else
       SECTION .rotext align=SECTION_ALIGN
 %endif
