--- lib/mp4v2/atom_avc1.cpp.orig   2007-07-24 02:19:41.000000000 -0400
+++ lib/mp4v2/atom_avc1.cpp   2007-07-24 02:19:44.000000000 -0400
@@ -41,6 +41,7 @@
	MP4StringProperty* pProp =
		new MP4StringProperty("compressorName");
	pProp->SetFixedLength(32);
+	pProp->SetCountedFormat(true);
	pProp->SetValue("AVC Coding");
	AddProperty(pProp); /* 6 */

--- lib/mp4v2/mp4property.cpp.orig   2007-07-24 02:20:32.000000000 -0400
+++ lib/mp4v2/mp4property.cpp   2007-07-24 02:21:52.000000000 -0400
@@ -391,6 +391,10 @@
	if (m_useCountedFormat) {
		pFile->WriteCountedString(m_values[index],
			(m_useUnicode ? 2 : 1), m_useExpandedCount);
+	if (m_fixedLength) {
+		pFile->WriteBytes((u_int8_t*)m_values[index],
+			m_fixedLength - strlen(m_values[index]) - 1);
+		}
	} else if (m_fixedLength) {
		pFile->WriteBytes((u_int8_t*)m_values[index], m_fixedLength);
	} else {
    
