diff -Naur x264/configure x264-patched/configure
--- x264/configure	2007-01-14 12:30:22.000000000 -0500
+++ x264-patched/configure	2007-01-14 13:37:18.078125000 -0500
@@ -113,8 +113,8 @@
     ;;
   cygwin*)
     SYS="CYGWIN"
-    CFLAGS="$CFLAGS -mno-cygwin"
-    LDFLAGS="$LDFLAGS -mno-cygwin"
+    #CFLAGS="$CFLAGS -mno-cygwin"
+    #LDFLAGS="$LDFLAGS -mno-cygwin"
     EXE=".exe"
     DEVNULL="NUL"
     ;;
