--- ffmpeg.base/configure	Wed Oct 10 15:39:57 2007
+++ ffmpeg/configure	Wed Oct 10 15:51:44 2007
@@ -889,7 +889,7 @@
     source_path_used="no"
 else
     source_path="`cd \"$source_path\"; pwd`"
-    echo "$source_path" | grep -q '[[:blank:]]' &&
+    echo "$source_path" | /usr/xpg4/bin/grep -q '[[:blank:]]' &&
       die "Out of tree builds are impossible with whitespace in source path."
 fi
 
@@ -993,7 +993,7 @@
   ;;
   --enable-?*|--disable-?*)
   eval `echo "$opt" | sed 's/--/action=/;s/-/ option=/;s/-/_/g'`
-  echo "$CMDLINE_SELECT" | grep -q "^ *$option\$" || die_unknown $opt
+  echo "$CMDLINE_SELECT" | /usr/xpg4/bin/grep -q "^ *$option\$" || die_unknown $opt
   $action $option
   ;;
   --list-*)
@@ -1368,7 +1368,7 @@
 fi
 
 gnu_make(){
-    $1 --version 2>&1 | grep -q GNU
+    $1 --version 2>&1 | /usr/xpg4/bin/grep -q GNU
 }
 
 if ! gnu_make $make; then
