diff -ruN xvidcore-orig/build/generic/configure xvidcore/build/generic/configure
--- xvidcore-orig/build/generic/configure	2006-11-01 03:39:36.000000000 -0700
+++ xvidcore/build/generic/configure	2008-01-30 20:19:48.000000000 -0700
@@ -4016,7 +4016,7 @@
        if test "$ac_nasm" = "yes" ; then
                                             echo "$as_me:$LINENO: checking for nasm patch version" >&5
 echo $ECHO_N "checking for nasm patch version... $ECHO_C" >&6
-           nasm_patch=`$nasm_prog -r | cut -d '.' -f 3 | cut -d ' ' -f 1`
+           nasm_patch=`($nasm_prog -r || $nasm_prog -v) | cut -d '.' -f 3 | cut -d ' ' -f 1`
            if test -z $nasm_patch ; then
-              nasm_patch=-1
+              nasm_patch=45
            fi
