diff -Naur libquicktime/configure libquicktime_patched/configure
--- libquicktime/configure	2006-08-03 01:02:25.000000000 +0200
+++ libquicktime_patched/configure	2007-02-21 16:16:22.000000000 +0100
@@ -25564,7 +25564,7 @@
 
 
 
-have_libavcodec=false
+have_libavcodec="true"
 
 AVCODEC_BUILD="3277056"
 
@@ -25656,6 +25656,8 @@
 fi
 rm -f core *.core gmon.out bb.out conftest$ac_exeext conftest.$ac_objext conftest.$ac_ext
 fi
+have_libavcodec="true"
+avcodec_ok="true"
 CFLAGS="$CFLAGS_save"
 LIBS="$LIBS_save"
 
@@ -26048,7 +26050,7 @@
 
 
 
-have_faac="false"
+have_faac="true"
 FAAC_REQUIRED="1.24"
 
 
@@ -26134,6 +26136,7 @@
 
 
 
+have_faac="true"
 
 if test x$have_faac = xtrue; then
   HAVE_FAAC_TRUE=
@@ -26261,7 +26264,7 @@
 
 
 
-have_x264="false"
+have_x264="true"
 
 if test $have_gpl = "true"; then
 X264_REQUIRED="0.48"
@@ -26321,6 +26324,7 @@
 
 fi
 
+have_x264="true"
 
 if test x$have_x264 = xtrue; then
   HAVE_X264_TRUE=
diff -Naur libquicktime/plugins/x264/x264.c libquicktime_patched/plugins/x264/x264.c
--- libquicktime/plugins/x264/x264.c	2006-07-25 13:35:15.000000000 +0200
+++ libquicktime_patched/plugins/x264/x264.c	2007-02-21 16:14:11.000000000 +0100
@@ -493,7 +493,7 @@
       {
       /* Force ABR */
       codec->params.rc.i_rc_method = X264_RC_ABR;
-      codec->params.rc.i_rf_constant = 0;
+      // codec->params.rc.i_rf_constant = 0;
       if(codec->pass == 1)
         {
         /* Strings will be made private by x264 */
@@ -655,7 +655,7 @@
   INTPARAM("x264_i_bitrate", codec->params.rc.i_bitrate);
   
   INTPARAM("x264_i_qp_constant", codec->params.rc.i_qp_constant);
-  INTPARAM("x264_i_rf_constant", codec->params.rc.i_rf_constant);
+  // INTPARAM("x264_i_rf_constant", codec->params.rc.i_rf_constant);
   INTPARAM("x264_i_qp_min", codec->params.rc.i_qp_min);
   INTPARAM("x264_i_qp_max", codec->params.rc.i_qp_max);
   INTPARAM("x264_i_qp_step", codec->params.rc.i_qp_step);
