diff -ru x264/Makefile x264.solaris/Makefile
--- x264/Makefile	Mon Jun  9 01:36:08 2008
+++ x264.solaris/Makefile	Thu Jun 19 11:54:53 2008
@@ -150,8 +150,10 @@
 	$(MAKE) -C gtk distclean
 
 install: x264$(EXE) $(SONAME)
-	install -d $(DESTDIR)$(bindir) $(DESTDIR)$(includedir)
-	install -d $(DESTDIR)$(libdir) $(DESTDIR)$(libdir)/pkgconfig
+	install -d $(DESTDIR)$(bindir) 
+	install -d $(DESTDIR)$(includedir)
+	install -d $(DESTDIR)$(libdir) 
+	install -d $(DESTDIR)$(libdir)/pkgconfig
 	install -m 644 x264.h $(DESTDIR)$(includedir)
 	install -m 644 libx264.a $(DESTDIR)$(libdir)
 	install -m 644 x264.pc $(DESTDIR)$(libdir)/pkgconfig
diff -ru x264/configure x264.solaris/configure
--- x264/configure	Mon Jun  9 01:36:08 2008
+++ x264.solaris/configure	Thu Jun 19 12:10:24 2008
@@ -1,4 +1,4 @@
-#! /bin/sh
+#! /usr/xpg4/bin/sh
 
 if test x"$1" = x"-h" -o x"$1" = x"--help" ; then
 
@@ -73,7 +73,7 @@
 # check whether 'echo -n' works as expected, otherwise try printf
 if [ "x`echo -n houba`" = xhouba ]
 then
-  ECHON="echo -n"
+  ECHON="/usr/ucb/echo -n"
 elif [ "x`printf houba`" = xhouba ]
 then
   ECHON="printf"
@@ -319,8 +319,10 @@
 
 echo "unsigned int endian = 'B' << 24 | 'I' << 16 | 'G' << 8 | 'E';" > conftest.c
 $CC $CFLAGS conftest.c -c -o conftest.o 2>$DEVNULL || die "endian test failed"
-grep -q BIGE conftest.o && CFLAGS="$CFLAGS -DWORDS_BIGENDIAN"
+/usr/bin/tr -cd "BIGE" < conftest.o | /usr/xpg4/bin/grep -q BIGE && CFLAGS="$CFLAGS -DWORDS_BIGENDIAN"
 
+echo $CFLAGS
+
 # autodetect options that weren't forced nor disabled
 
 libpthread=""
diff -ru x264/version.sh x264.solaris/version.sh
--- x264/version.sh	Mon Jun  9 01:36:08 2008
+++ x264.solaris/version.sh	Thu Jun 19 11:56:19 2008
@@ -1,7 +1,7 @@
 #!/bin/sh
 git-rev-list HEAD | sort > config.git-hash
 LOCALVER=`wc -l config.git-hash | awk '{print $1}'`
-if [ $LOCALVER \> 1 ] ; then
+if [ $LOCALVER -gt 1 ] ; then
     VER=`git-rev-list origin/master | sort | join config.git-hash - | wc -l | awk '{print $1}'`
     if [ $VER != $LOCALVER ] ; then
         VER="$VER+$(($LOCALVER-$VER))"
diff -ru x264/common/osdep.h x264.solaris/common/osdep.h
--- x264/common/osdep.h	Mon Jun  9 01:36:08 2008
+++ x264.solaris/common/osdep.h	Thu Jun 19 10:52:56 2008
@@ -48,6 +48,10 @@
 #define X264_VERSION "" // no configure script for msvc
 #endif
 
+#if defined(SYS_SunOS)
+#define isfinite _finite
+#endif
+
 #ifdef SYS_OPENBSD
 #define isfinite finite
 #endif
