Index: libavformat/tcp.c
===================================================================
--- libavformat/tcp.c	(revision 9814)
+++ libavformat/tcp.c	(working copy)
@@ -88,11 +88,13 @@
                 break;
         }
 
+#ifndef __BEOS__
         /* test error */
         optlen = sizeof(ret);
         getsockopt (fd, SOL_SOCKET, SO_ERROR, &ret, &optlen);
         if (ret != 0)
             goto fail;
+#endif
     }
     s->fd = fd;
     return 0;
