diff --git a/libswscale/utils.c b/libswscale/utils.c
index f0a2b46..61506c6 100644
--- a/libswscale/utils.c
+++ b/libswscale/utils.c
@@ -531,7 +531,9 @@ static int initFilter(int16_t **outFilter, int32_t **filterPos,
     // FIXME try to align filterPos if possible
 
     // fix borders
-    if (is_horizontal) {
+    // This "horizontal" only check causes a subsequent assert when
+    // upscaling.
+    /*if (is_horizontal)*/ {
         for (i = 0; i < dstW; i++) {
             int j;
             if ((*filterPos)[i] < 0) {
