diff --git a/libavcodec/Makefile b/libavcodec/Makefile
index 3c4e2f8..fe84b8c 100644
--- a/libavcodec/Makefile
+++ b/libavcodec/Makefile
@@ -1,7 +1,7 @@
 NAME = avcodec
 FFLIBS = avutil
 
-HEADERS = avcodec.h avfft.h dxva2.h opt.h vaapi.h vdpau.h version.h xvmc.h
+HEADERS = avcodec.h avfft.h dxva2.h opt.h vaapi.h vdpau.h version.h xvmc.h audioconvert.h
 
 OBJS = allcodecs.o                                                      \
        audioconvert.o                                                   \
