diff --git a/libavcodec/pngdec.c b/libavcodec/pngdec.c
index 7477f67..9e8e29b 100644
--- a/libavcodec/pngdec.c
+++ b/libavcodec/pngdec.c
@@ -599,6 +599,18 @@ static int decode_frame(AVCodecContext *avctx,
         }
     }
  exit_loop:
+/* ffmpeg-r18421 introduced support for CorePNG p-frames which
+ * breaks raw png sequences created by QuickTime Pro.
+ * since only the first png appears to be marked as PKT_FLAG_KEY
+ * it means either QuickTime Pro is encoding incorrectly, or
+ * this code block needs to find other criteria.
+ *
+ * As of ffmpeg-r19079 this patch should no longer be required.
+ * It offers a similar solution; forces code block to be skipped.
+ *
+ * --kb
+ */
+#if 0
      /* handle p-frames only if a predecessor frame is available */
      if(s->last_picture->data[0] != NULL) {
          if(!(avpkt->flags & AV_PKT_FLAG_KEY)) {
@@ -615,6 +627,7 @@ static int decode_frame(AVCodecContext *avctx,
             }
         }
     }
+#endif
 
     *picture= *s->current_picture;
     *data_size = sizeof(AVFrame);
