commit d995f0a1550c9148140eab34c7d770e90925c004
Author: Luca Barbato <lu_zero@gentoo.org>
Date:   Sat Sep 13 00:26:21 2014 +0200

    log: Unbreak no-tty support on 256color terminals

diff --git a/libavutil/log.c b/libavutil/log.c
index 5a8f293..d38e40b 100644
--- a/libavutil/log.c
+++ b/libavutil/log.c
@@ -75,7 +75,8 @@ static void check_color_terminal(void)
     char *term = getenv("TERM");
     use_color = !getenv("NO_COLOR") && !getenv("AV_LOG_FORCE_NOCOLOR") &&
                 (getenv("TERM") && isatty(2) || getenv("AV_LOG_FORCE_COLOR"));
-    use_color += term && strstr(term, "256color");
+    if (use_color)
+        use_color += term && strstr(term, "256color");
 #else
     use_color = getenv("AV_LOG_FORCE_COLOR") && !getenv("NO_COLOR") &&
                !getenv("AV_LOG_FORCE_NOCOLOR");


