diff --git a/libavutil/x86/float_dsp_init.c b/libavutil/x86/float_dsp_init.c
index d259a36..3729d91 100644
--- a/libavutil/x86/float_dsp_init.c
+++ b/libavutil/x86/float_dsp_init.c
@@ -38,11 +38,13 @@ void ff_float_dsp_init_x86(AVFloatDSPContext *fdsp)
 
     if (mm_flags & AV_CPU_FLAG_SSE && HAVE_SSE) {
         fdsp->vector_fmul = ff_vector_fmul_sse;
-        fdsp->vector_fmac_scalar = ff_vector_fmac_scalar_sse;
+        // buggy under 64-bit Windows: http://bugzilla.libav.org/show_bug.cgi?id=336
+        // fdsp->vector_fmac_scalar = ff_vector_fmac_scalar_sse;
     }
     if (mm_flags & AV_CPU_FLAG_AVX && HAVE_AVX) {
         fdsp->vector_fmul = ff_vector_fmul_avx;
-        fdsp->vector_fmac_scalar = ff_vector_fmac_scalar_avx;
+        // buggy under 64-bit Windows: http://bugzilla.libav.org/show_bug.cgi?id=336
+        // fdsp->vector_fmac_scalar = ff_vector_fmac_scalar_avx;
     }
 #endif
 }
