commit f94371b1db89bf65173a763a49175ce549466602
Author: Michael Niedermayer <michaelni@gmx.at>
Date:   Sat Mar 29 15:56:03 2014 +0100

    mov: Unbreak sbtl writing

diff --git a/libavformat/movenc.c b/libavformat/movenc.c
index a6a55d5..889f15c 100644
--- a/libavformat/movenc.c
+++ b/libavformat/movenc.c
@@ -1312,9 +1312,13 @@ static int mov_write_hdlr_tag(AVIOContext *pb, MOVTrack *track)
             hdlr_type = "soun";
             descr     = "SoundHandler";
         } else if (track->enc->codec_type == AVMEDIA_TYPE_SUBTITLE) {
-            if (track->tag == MKTAG('t','x','3','g')) hdlr_type = "sbtl";
-            if (track->tag == MKTAG('m','p','4','s')) hdlr_type = "subp";
-            else                                      hdlr_type = "text";
+            if (track->tag == MKTAG('t','x','3','g')) {
+                hdlr_type = "sbtl";
+            } else if (track->tag == MKTAG('m','p','4','s')) {
+                hdlr_type = "subp";
+            } else {
+                hdlr_type = "text";
+            }
             descr = "SubtitleHandler";
         } else if (track->enc->codec_tag == MKTAG('r','t','p',' ')) {
             hdlr_type = "hint";
