diff --git a/libavcodec/h264_parser.c b/libavcodec/h264_parser.c
index 145dce3..b9c47ab 100644
--- a/libavcodec/h264_parser.c
+++ b/libavcodec/h264_parser.c
@@ -474,7 +474,8 @@ static int h264_split(AVCodecContext *avctx,
          *  }
          */
         if ((state & 0xFFFFFF00) == 0x100 && (state & 0xFFFFFF1F) != 0x107 &&
-            (state & 0xFFFFFF1F) != 0x108 && (state & 0xFFFFFF1F) != 0x109) {
+            (state & 0xFFFFFF1F) != 0x108 && (state & 0xFFFFFF1F) != 0x109 &&
+            (state & 0xFFFFFF1F) != 0x10d) {
             if (has_sps) {
                 while (i > 4 && buf[i - 5] == 0)
                     i--;
