From 1bac8a0d52ec58b914f308fe0b00b1380d87c1a6 Mon Sep 17 00:00:00 2001
From: Vladyslav Sosunovych <vladyslav.sosunovych@intel.com>
Date: Tue, 25 Jul 2023 13:04:29 +0200
Subject: [PATCH] lavfi/vf_vpp_qsv: fix duration in pass-through mode

---
 libavfilter/vf_vpp_qsv.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/libavfilter/vf_vpp_qsv.c b/libavfilter/vf_vpp_qsv.c
index b807fa5e63..aa289538f9 100644
--- a/libavfilter/vf_vpp_qsv.c
+++ b/libavfilter/vf_vpp_qsv.c
@@ -767,6 +767,11 @@ static int activate(AVFilterContext *ctx)
             if (in->pts != AV_NOPTS_VALUE)
                 in->pts = av_rescale_q(in->pts, inlink->time_base, outlink->time_base);
 
+            if (outlink->frame_rate.num && outlink->frame_rate.den)
+                in->duration = av_rescale_q(1, av_inv_q(outlink->frame_rate), outlink->time_base);
+            else
+                in->duration = 0;
+
             ret = ff_filter_frame(outlink, in);
             if (ret < 0)
                 return ret;
-- 
2.40.1

