From 18f1d7d36534f059c3a5709ff1facda6a26377bd Mon Sep 17 00:00:00 2001
From: Damiano Galassi <damiog@gmail.com>
Date: Fri, 22 Aug 2025 17:01:26 +0200
Subject: [PATCH 22/22] movenc: ensure chapters track extradata is not null and
 populated

---
 libavformat/movenc.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/libavformat/movenc.c b/libavformat/movenc.c
index 4136a5d623..cb36f6d760 100644
--- a/libavformat/movenc.c
+++ b/libavformat/movenc.c
@@ -7592,6 +7592,12 @@ static int mov_create_chapter_track(AVFormatContext *s, int tracknum)
         return ret;
     memcpy(track->par->extradata, stub_header, sizeof(stub_header));
 
+    if (track->extradata == NULL) {
+        track->stsd_count     = 1;
+        track->extradata      = av_calloc(1, sizeof(*track->extradata));
+        track->extradata_size = av_calloc(1, sizeof(*track->extradata_size));
+    }
+
     track->extradata[0] = av_memdup(stub_header, sizeof(stub_header));
     if (!track->extradata[0])
         return AVERROR(ENOMEM);
-- 
2.39.5 (Apple Git-154)

