From 1f7ac8d10a7a32ae3d36a473c015b31a8aedc7ea Mon Sep 17 00:00:00 2001
From: Damiano Galassi <damiog@gmail.com>
Date: Mon, 14 Aug 2023 07:16:39 +0200
Subject: [PATCH] videotoolbox: disable H.264 10-bit on Intel macOS, it's not
 decoded properly.

---
 libavcodec/videotoolbox.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/libavcodec/videotoolbox.c b/libavcodec/videotoolbox.c
index e42fea6..76257f8 100644
--- a/libavcodec/videotoolbox.c
+++ b/libavcodec/videotoolbox.c
@@ -911,6 +911,15 @@ static int videotoolbox_start(AVCodecContext *avctx)
         break;
     }
 
+#if ARCH_X86_64
+    if (avctx->codec_id == AV_CODEC_ID_H264 &&
+        avctx->sw_pix_fmt == AV_PIX_FMT_YUV420P10)
+    {
+        // 10-bit H.264 is not supporte on x86_64
+        return AVERROR(ENOSYS);
+    }
+#endif
+
 #if defined(MAC_OS_X_VERSION_10_9) && !TARGET_OS_IPHONE && (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_9) && AV_HAS_BUILTIN(__builtin_available)
     if (avctx->codec_id == AV_CODEC_ID_PRORES) {
         if (__builtin_available(macOS 10.9, *)) {
-- 
2.39.3 (Apple Git-145)

