From c7e60e75926bf05203e45720b479a7b0c1bf866d Mon Sep 17 00:00:00 2001
From: Damiano Galassi <damiog@gmail.com>
Date: Thu, 12 Feb 2026 09:37:21 +0100
Subject: [PATCH] movenc: always use version 2 descriptor for pcm

---
 libavformat/movenc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/libavformat/movenc.c b/libavformat/movenc.c
index 5e80fa57d6..07d19ad9e2 100644
--- a/libavformat/movenc.c
+++ b/libavformat/movenc.c
@@ -1353,7 +1353,8 @@ static int mov_write_audio_tag(AVFormatContext *s, AVIOContext *pb, MOVMuxContex
     int ret = 0;
 
     if (track->mode == MODE_MOV) {
-        if (track->timescale > UINT16_MAX || !track->par->ch_layout.nb_channels) {
+        if (track->timescale > UINT16_MAX || !track->par->ch_layout.nb_channels ||
+            mov_get_lpcm_flags(track->par->codec_id)) {
             if (mov_get_lpcm_flags(track->par->codec_id))
                 tag = AV_RL32("lpcm");
             version = 2;
-- 
2.50.1 (Apple Git-155)

