From 703e4a5cea1a75d384cb7b29cca398767312c08a Mon Sep 17 00:00:00 2001
From: Nomis101 <Nomis101@web.de>
Date: Sun, 3 Jul 2022 19:39:44 +0200
Subject: [PATCH] enable TrueHD for MP4

Signed-off-by: Nomis101 <Nomis101@web.de>
---
 libavformat/movenc.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/libavformat/movenc.c b/libavformat/movenc.c
index 4c868919ae..f811f9a927 100644
--- a/libavformat/movenc.c
+++ b/libavformat/movenc.c
@@ -6848,7 +6848,6 @@ static int mov_init(AVFormatContext *s)
                 }
             }
             if (track->par->codec_id == AV_CODEC_ID_FLAC ||
-                track->par->codec_id == AV_CODEC_ID_TRUEHD ||
                 track->par->codec_id == AV_CODEC_ID_OPUS) {
                 if (track->mode != MODE_MP4) {
                     av_log(s, AV_LOG_ERROR, "%s only supported in MP4.\n", avcodec_get_name(track->par->codec_id));
-- 
2.32.0 (Apple Git-132)

