From 000dc4f0e502e05c3381916494cbf6b1eea2c932 Mon Sep 17 00:00:00 2001
From: OvchinnikovDmitrii <ovchinnikov.dmitrii@gmail.com>
Date: Tue, 5 Jul 2022 10:36:12 +0200
Subject: [avcodec/amfenc: 10 bit support v2 1/3] amfenc: Update the min version to
 1.4.23.0 for AMF SDK.

---
 configure | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure b/configure
index 9e51abd0d3..a644e17c95 100755
--- a/configure
+++ b/configure
@@ -7038,7 +7038,7 @@ fi
 
 enabled amf &&
     check_cpp_condition amf "AMF/core/Version.h" \
-        "(AMF_VERSION_MAJOR << 48 | AMF_VERSION_MINOR << 32 | AMF_VERSION_RELEASE << 16 | AMF_VERSION_BUILD_NUM) >= 0x0001000400090000"
+        "(AMF_VERSION_MAJOR << 48 | AMF_VERSION_MINOR << 32 | AMF_VERSION_RELEASE << 16 | AMF_VERSION_BUILD_NUM) >= 0x0001000400170000"
 
 # Funny iconv installations are not unusual, so check it after all flags have been set
 if enabled libc_iconv; then
-- 
2.30.0.windows.2

