diff --git a/configure b/configure
index 2072c5e..7b52128 100755
--- a/configure
+++ b/configure
@@ -3402,7 +3402,7 @@ av1_qsv_encoder_deps="libvpl"
 av1_amf_encoder_deps="amf"
 av1_vaapi_encoder_deps="VAEncPictureParameterBufferAV1"
 av1_vaapi_encoder_select="cbs_av1 vaapi_encode"
-vvc_qsv_decoder_select="qsvdec"
+vvc_qsv_decoder_select="vvc_mp4toannexb_bsf qsvdec"
 
 # parsers
 aac_parser_select="adts_header mpeg4audio"

diff --git a/libavcodec/qsvdec.c b/libavcodec/qsvdec.c
index 56e131b..9ad3269 100644
--- a/libavcodec/qsvdec.c
+++ b/libavcodec/qsvdec.c
@@ -1302,5 +1302,5 @@ DEFINE_QSV_DECODER(av1, AV1, NULL)
 #endif
 
 #if CONFIG_VVC_QSV_DECODER
-DEFINE_QSV_DECODER(vvc, VVC, NULL)
+DEFINE_QSV_DECODER(vvc, VVC, "vvc_mp4toannexb")
 #endif
