diff --git a/libavcodec/videotoolboxenc.c b/libavcodec/videotoolboxenc.c
index cc08cf6..1d83408 100644
--- a/libavcodec/videotoolboxenc.c
+++ b/libavcodec/videotoolboxenc.c
@@ -940,21 +940,17 @@ static int get_cv_transfer_function(AVCodecContext *avctx,
             *transfer_fnc = kCVImageBufferTransferFunction_SMPTE_240M_1995;
             break;
 
-#if HAVE_KCVIMAGEBUFFERTRANSFERFUNCTION_SMPTE_ST_2084_PQ
         case AVCOL_TRC_SMPTE2084:
-            *transfer_fnc = kCVImageBufferTransferFunction_SMPTE_ST_2084_PQ;
+            *transfer_fnc = CFSTR("SMPTE_ST_2084_PQ");
             break;
-#endif
-#if HAVE_KCVIMAGEBUFFERTRANSFERFUNCTION_LINEAR
+
         case AVCOL_TRC_LINEAR:
-            *transfer_fnc = kCVImageBufferTransferFunction_Linear;
+            *transfer_fnc = CFSTR("Linear");
             break;
-#endif
-#if HAVE_KCVIMAGEBUFFERTRANSFERFUNCTION_ITU_R_2100_HLG
+
         case AVCOL_TRC_ARIB_STD_B67:
-            *transfer_fnc = kCVImageBufferTransferFunction_ITU_R_2100_HLG;
+            *transfer_fnc = CFSTR("ITU_R_2100_HLG");
             break;
-#endif
 
         case AVCOL_TRC_GAMMA22:
             gamma = 2.2;
