Index: configure
===================================================================
--- ffmpeg.orig/configure	(revision 20594)
+++ ffmpeg/configure	(working copy)
@@ -54,6 +54,9 @@
     exit 1
 fi
 
+# Solaris must use the xpg4 version of grep
+PATH=/usr/xpg4/bin:$PATH
+
 show_help(){
 cat <<EOF
 Usage: configure [options]
@@ -2171,7 +2174,7 @@
 check_cc <<EOF || die "endian test failed"
 unsigned int endian = 'B' << 24 | 'I' << 16 | 'G' << 8 | 'E';
 EOF
-od -A n -t x1 $TMPO | grep -q '42 *49 *47 *45' && enable bigendian
+/usr/bin/tr -cd "BIGE" < $TMPO | grep -q 'B *I *G *E' && enable bigendian
 
 if enabled arm; then
 
