diff --git a/configure b/configure
index aa31ea0..773e6bf 100755
--- a/configure
+++ b/configure
@@ -3167,7 +3167,7 @@ EOF
 check_cc <<EOF || die "endian test failed"
 unsigned int endian = 'B' << 24 | 'I' << 16 | 'G' << 8 | 'E';
 EOF
-od -t x1 $TMPO | grep -q '42 *49 *47 *45' && enable bigendian
+/usr/bin/tr -cd "BIGE" < $TMPO | grep -q 'B *I *G *E' && enable bigendian
 
 if enabled alpha; then
 
