diff --git a/configure b/configure
index 5085333..3d85303 100755
--- a/configure
+++ b/configure
@@ -3808,7 +3808,7 @@ EOF
 check_cc <<EOF || die "endian test failed"
 unsigned int endian = 'B' << 24 | 'I' << 16 | 'G' << 8 | 'E';
 EOF
-od -t x1 $TMPO | grep -q '42 *49 *47 *45' && enable bigendian
+/usr/bin/tr -cd "BIGE" < $TMPO | grep -q 'B *I *G *E' && enable bigendian
 
 check_gas() {
     log "check_gas using '$as' as AS"
