diff --git a/configure b/configure
index e9fc08c..86ebc37 100755
--- a/configure
+++ b/configure
@@ -3589,7 +3589,7 @@ EOF
 check_cc <<EOF || die "endian test failed"
 unsigned int endian = 'B' << 24 | 'I' << 16 | 'G' << 8 | 'E';
 EOF
-od -t x1 $TMPO | grep -q '42 *49 *47 *45' && enable bigendian
+/usr/bin/tr -cd "BIGE" < $TMPO | grep -q 'B *I *G *E' && enable bigendian
 
 check_inline_asm inline_asm_labels '"1:\n"'
 
