Index: ffmpeg/libavcodec/x86/dsputil_mmx.c
===================================================================
--- ffmpeg/libavcodec/x86/dsputil_mmx.c	(revision 14508)
+++ ffmpeg/libavcodec/x86/dsputil_mmx.c	(working copy)
@@ -2907,6 +2907,8 @@
                 c->vp6_filter_diag4 = ff_vp6_filter_diag4_sse2;
             }
         }
+#undef  H264_QPEL_FUNCS
+#define H264_QPEL_FUNCS(x, y, CPU) 
 #if HAVE_SSSE3
         if(mm_flags & MM_SSSE3){
             H264_QPEL_FUNCS(1, 0, ssse3);
