diff --git a/libavutil/mem.c b/libavutil/mem.c
index bf1a542..a32d400 100644
--- a/libavutil/mem.c
+++ b/libavutil/mem.c
@@ -82,6 +82,11 @@ void *av_malloc(size_t size)
     ptr = (char*)ptr + diff;
     ((char*)ptr)[-1]= diff;
 #elif HAVE_POSIX_MEMALIGN
+#if defined(__APPLE__)
+    if (size == 0)
+        ptr = NULL;
+    else
+#endif
     if (posix_memalign(&ptr,32,size))
         ptr = NULL;
 #elif HAVE_MEMALIGN
