From e1fd1d4a36637010e455db28546313d3b2939679 Mon Sep 17 00:00:00 2001
From: sekrit-twc <noreply@example.com>
Date: Fri, 16 Apr 2021 12:23:17 -0700
Subject: [PATCH] common: fix alignment for ARM

---
 src/zimg/common/align.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/zimg/common/align.h b/src/zimg/common/align.h
index a30468c..7de840a 100644
--- a/src/zimg/common/align.h
+++ b/src/zimg/common/align.h
@@ -11,6 +11,9 @@ namespace zimg {
 #ifdef ZIMG_X86
 constexpr int ALIGNMENT = 64;
 constexpr int ALIGNMENT_RELAXED = 32;
+#elif ZIMG_ARM
+constexpr int ALIGNMENT = 16;
+constexpr int ALIGNMENT_RELAXED = 16;
 #else
 constexpr int ALIGNMENT = sizeof(long double);
 constexpr int ALIGNMENT_RELAXED = sizeof(long double);
-- 
2.24.3 (Apple Git-128)

