diff -u fribidi-0.19.7/bin/Makefile.am.orig fribidi-0.19.7/bin/Makefile.am
--- fribidi-0.19.7/bin/Makefile.am.orig	2012-08-27 15:30:17.000000000 -0400
+++ fribidi-0.19.7/bin/Makefile.am	2015-09-17 22:57:45.000000000 -0400
@@ -1,6 +1,6 @@
 bin_PROGRAMS = fribidi
 
-noinst_PROGRAMS = fribidi-benchmark fribidi-bidi-types
+noinst_PROGRAMS = fribidi-bidi-types
 if FRIBIDI_CHARSETS
 noinst_PROGRAMS += fribidi-caprtl2utf8
 endif
