diff -ur fribidi-1.0.17.orig/Makefile.am fribidi-1.0.17/Makefile.am
--- fribidi-1.0.17.orig/Makefile.am 2026-09-20 07:51:31.000000000 -0400
+++ fribidi-1.0.17/Makefile.am 2026-09-20 10:15:31.000000000 -0500
@@ -4,7 +4,7 @@
 ACLOCAL_AMFLAGS = -I m4
 
 ## The order of subdirs is important, don't change without a reason.
-SUBDIRS = gen.tab lib bin doc test
+SUBDIRS = gen.tab lib
 
 EXTRA_DIST = autogen.sh ChangeLog.old \
 	meson.build meson_options.txt bin/meson.build doc/meson.build \
