diff --git a/lib/fribidi-common.h b/lib/fribidi-common.h
index d874d8b..9388a6f 100644
--- a/lib/fribidi-common.h
+++ b/lib/fribidi-common.h
@@ -50,9 +50,9 @@
 /* Windows is special and you cannot just define entry points unconditionally. */
 # elif defined(_WIN32) && ! defined(FRIBIDI_LIB_STATIC)
 #  ifdef FRIBIDI_BUILD
-#   define FRIBIDI_ENTRY __declspec(dllexport)
+#  define FRIBIDI_ENTRY
 #  else
-#   define FRIBIDI_ENTRY __declspec(dllimport)
+#  define FRIBIDI_ENTRY
 #  endif
 # else
 /* nothing else worked, give up and do nothing */
 