diff -Naur libdvdnav.orig/src/vm/vm.c libdvdnav/src/vm/vm.c
--- libdvdnav.orig/src/vm/vm.c	2009-03-13 18:28:22.000000000 -0700
+++ libdvdnav/src/vm/vm.c	2009-10-11 17:49:17.813221643 -0700
@@ -1866,6 +1866,13 @@
   default:
     abort();
   }
+  // get_MENU_PGCIT fails if h->pgci_ut == NULL.
+  // this happens with badly mastered discs.  Mac The Ripper happens
+  // to create such discs when in "Main Feature Extraction" mode.
+  if(pgcit == NULL && vm->vtsi != NULL) {
+    (vm->state).domain = VTS_DOMAIN;
+    pgcit = vm->vtsi->vts_pgcit;
+  }
 
   return pgcit;
 }
