--- nv-codec-headers-12.2.72.0/include/ffnvcodec/dynlink_loader.h
+++ nv-codec-headers-12.2.72.0/include/ffnvcodec/dynlink_loader.h
@@ -62,7 +62,7 @@

 #if !defined(FFNV_LOAD_FUNC) || !defined(FFNV_SYM_FUNC)
 # ifdef _WIN32
-#  define FFNV_LOAD_FUNC(path) LoadLibrary(TEXT(path))
+#  define FFNV_LOAD_FUNC(path) LoadLibraryExA(TEXT(path), NULL, LOAD_LIBRARY_SEARCH_SYSTEM32)
 #  define FFNV_SYM_FUNC(lib, sym) GetProcAddress((lib), (sym))
 #  define FFNV_FREE_FUNC(lib) FreeLibrary(lib)
 # else
