diff -ruN libopusenc-v0.3/libopusenc-uninstalled.pc.in libopusenc-v0.3-patched/libopusenc-uninstalled.pc.in
--- libopusenc-v0.3/libopusenc-uninstalled.pc.in	2025-12-14 18:00:00.000000000 +0100
+++ libopusenc-v0.3-patched/libopusenc-uninstalled.pc.in	2025-12-21 17:43:37.210934672 +0100
@@ -8,7 +8,7 @@
 Name: libopusenc uninstalled
 Description: High-level Opus encoding library (not installed)
 URL: https://opus-codec.org/
-Version: @PACKAGE_VERSION@
+Version: 0.3
 Requires.private: opus >= 1.1
 Conflicts:
 Libs: ${libdir}/libopusenc.la @lrintf_lib@
diff -ruN libopusenc-v0.3/libopusenc.pc.in libopusenc-v0.3-patched/libopusenc.pc.in
--- libopusenc-v0.3/libopusenc.pc.in	2025-12-14 18:00:00.000000000 +0100
+++ libopusenc-v0.3-patched/libopusenc.pc.in	2025-12-21 17:43:37.307218290 +0100
@@ -8,7 +8,7 @@
 Name: libopusenc
 Description: High-level Opus encoding library
 URL: https://opus-codec.org/
-Version: @PACKAGE_VERSION@
+Version: 0.3
 Requires.private: opus >= 1.1
 Conflicts:
 Libs: -L${libdir} -lopusenc
