--- libbluray-1.3.3/configure.ac.orig	Wed Oct 19 01:23:23 2022
+++ libbluray-1.3.3/configure.ac	Wed Oct 19 01:24:43 2022
@@ -128,6 +128,12 @@ AC_ARG_WITH([bdj-bootclasspath],
 AC_ARG_WITH([java9],
   [AS_HELP_STRING([--with-java9], [build with Java 9+ (drop support for Java <1.6) @<:@default=without@:>@])])
 
+AC_ARG_WITH([external-libudfread],
+  [AS_HELP_STRING([--without-external-libudfread],
+  [build without external libudfread @<:@default=with@:>@])],
+  [with_external_libudfread=$withwal],
+  [with_external_libudfread=yes])
+
 dnl required programs
 AC_PROG_CC
 AM_PROG_CC_C_O
@@ -309,11 +315,14 @@ AM_CONDITIONAL([BDJ_J2SE], [ test x"$BDJ_TYPE" != x"j2
 dnl bootclasspath
 AC_SUBST(BDJ_BOOTCLASSPATH)
 
-PKG_CHECK_MODULES([LIBUDFREAD], [libudfread >= 1.1.0],
-    [with_libudfread=yes
-     AC_DEFINE([HAVE_LIBUDFREAD], [1], [Define to 1 if external libudfread is to be used])
-     PACKAGES="$PACKAGES libudfread >= 1.1.0"
-    ],[
+AS_IF([test "x$with_external_libudfread" = "xyes"], [
+    PKG_CHECK_MODULES([LIBUDFREAD], [libudfread >= 1.1.0],
+        [with_libudfread=yes
+         AC_DEFINE([HAVE_LIBUDFREAD], [1], [Define to 1 if external libudfread is to be used])
+         PACKAGES="$PACKAGES libudfread >= 1.1.0"
+        ],[break;])
+])
+AS_IF([test "x$with_libudfread" != "xyes"], [
      dnl udf support (using git submodule)
      with_libudfread=no
      AS_IF([test ! -f "${srcdir}/contrib/libudfread/src/udfread.h"], [AC_MSG_ERROR("libudfread source tree not found")])
--- libbluray-1.3.3/configure.orig	Wed Oct 19 01:31:34 2022
+++ libbluray-1.3.3/configure	Wed Oct 19 01:31:47 2022
@@ -882,6 +882,7 @@ with_fontconfig
 with_bdj_type
 with_bdj_bootclasspath
 with_java9
+with_external_libudfread
 enable_shared
 enable_static
 with_pic
@@ -1597,6 +1598,8 @@ Optional Packages:
                           Default is empty.
   --with-java9            build with Java 9+ (drop support for Java <1.6)
                           [default=without]
+  --without-external-libudfread
+                          build without external libudfread [default=with]
   --with-pic[=PKGS]       try to use only PIC/non-PIC objects [default=use
                           both]
   --with-aix-soname=aix|svr4|both
@@ -5028,6 +5031,16 @@ then :
 fi
 
 
+
+# Check whether --with-external-libudfread was given.
+if test ${with_external_libudfread+y}
+then :
+  withval=$with_external_libudfread; with_external_libudfread=$withwal
+else $as_nop
+  with_external_libudfread=yes
+fi
+
+
 ac_ext=c
 ac_cpp='$CPP $CPPFLAGS'
 ac_compile='$CC -c $CFLAGS $CPPFLAGS conftest.$ac_ext >&5'
@@ -15847,7 +15860,10 @@ fi
 
 
 
+if test "x$with_external_libudfread" = "xyes"
+then :
 
+
 pkg_failed=no
 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for LIBUDFREAD" >&5
 printf %s "checking for LIBUDFREAD... " >&6; }
@@ -15904,44 +15920,28 @@ fi
 	# Put the nasty error message in config.log where it belongs
 	echo "$LIBUDFREAD_PKG_ERRORS" >&5
 
+	break;
+elif test $pkg_failed = untried; then
+     	{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: no" >&5
+printf "%s\n" "no" >&6; }
+	break;
+else
+	LIBUDFREAD_CFLAGS=$pkg_cv_LIBUDFREAD_CFLAGS
+	LIBUDFREAD_LIBS=$pkg_cv_LIBUDFREAD_LIBS
+        { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: yes" >&5
+printf "%s\n" "yes" >&6; }
+	with_libudfread=yes
 
-          with_libudfread=no
-     if test ! -f "${srcdir}/contrib/libudfread/src/udfread.h"
-then :
-  as_fn_error $? "\"libudfread source tree not found\"" "$LINENO" 5
-fi
-     ac_fn_c_check_header_compile "$LINENO" "unistd.h" "ac_cv_header_unistd_h" "$ac_includes_default"
-if test "x$ac_cv_header_unistd_h" = xyes
-then :
-  printf "%s\n" "#define HAVE_UNISTD_H 1" >>confdefs.h
+printf "%s\n" "#define HAVE_LIBUDFREAD 1" >>confdefs.h
 
-fi
-ac_fn_c_check_header_compile "$LINENO" "fcntl.h" "ac_cv_header_fcntl_h" "$ac_includes_default"
-if test "x$ac_cv_header_fcntl_h" = xyes
-then :
-  printf "%s\n" "#define HAVE_FCNTL_H 1" >>confdefs.h
+         PACKAGES="$PACKAGES libudfread >= 1.1.0"
 
 fi
 
-     if test "${SYS}" != "mingw32"
+fi
+if test "x$with_libudfread" != "xyes"
 then :
 
-         ac_fn_c_check_func "$LINENO" "pread" "ac_cv_func_pread"
-if test "x$ac_cv_func_pread" = xyes
-then :
-
-else $as_nop
-
-printf "%s\n" "#define NEED_PREAD_IMPL 1" >>confdefs.h
-
-fi
-
-fi
-
-elif test $pkg_failed = untried; then
-     	{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: no" >&5
-printf "%s\n" "no" >&6; }
-
           with_libudfread=no
      if test ! -f "${srcdir}/contrib/libudfread/src/udfread.h"
 then :
@@ -15975,17 +15975,6 @@ fi
 
 fi
 
-else
-	LIBUDFREAD_CFLAGS=$pkg_cv_LIBUDFREAD_CFLAGS
-	LIBUDFREAD_LIBS=$pkg_cv_LIBUDFREAD_LIBS
-        { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: yes" >&5
-printf "%s\n" "yes" >&6; }
-	with_libudfread=yes
-
-printf "%s\n" "#define HAVE_LIBUDFREAD 1" >>confdefs.h
-
-     PACKAGES="$PACKAGES libudfread >= 1.1.0"
-
 fi
  if  test x"$with_libudfread" = x"yes" ; then
   HAVE_LIBUDFREAD_TRUE=
@@ -20485,5 +20474,4 @@ fi
 echo "  Metadata support (libxml2):    $with_libxml2"
 echo "  External libudfread:           $with_libudfread"
 echo "  Build examples:                $use_examples"
-
 
