From 1a113167733091fe325152a5ece4d57ee1b89dc0 Mon Sep 17 00:00:00 2001
From: John Stebbins <stebbins@jetheaddev.com>
Date: Wed, 18 May 2016 15:40:17 -0600
Subject: [PATCH] expose clip_id in BLURAY_CLIP_INFO

---
 src/libbluray/bluray.c | 1 +
 src/libbluray/bluray.h | 1 +
 2 files changed, 2 insertions(+)

diff --git a/src/libbluray/bluray.c b/src/libbluray/bluray.c
index b7598b9..3478ade 100644
--- a/src/libbluray/bluray.c
+++ b/src/libbluray/bluray.c
@@ -2602,6 +2602,7 @@ static BLURAY_TITLE_INFO* _fill_title_info(NAV_TITLE* title, uint32_t title_idx,
         BLURAY_CLIP_INFO *ci = &title_info->clips[ii];
         NAV_CLIP *nc = &title->clip_list.clip[ii];
 
+        memcpy(ci->clip_id, pi->clip->clip_id, sizeof(ci->clip_id));
         ci->pkt_count = nc->end_pkt - nc->start_pkt;
         ci->start_time = (uint64_t)nc->title_time * 2;
         ci->in_time = (uint64_t)pi->in_time * 2;
diff --git a/src/libbluray/bluray.h b/src/libbluray/bluray.h
index 3cd7118..177b3d4 100644
--- a/src/libbluray/bluray.h
+++ b/src/libbluray/bluray.h
@@ -225,6 +225,7 @@ typedef struct bd_stream_info {
 } BLURAY_STREAM_INFO;
 
 typedef struct bd_clip {
+    char               clip_id[6];
     uint32_t           pkt_count;
     uint8_t            still_mode;
     uint16_t           still_time;  /* seconds */
-- 
2.5.5

