--- libbluray-0.2.3/src/Makefile.am.bak	Fri Aug 17 07:27:32 2012
+++ libbluray-0.2.3/src/Makefile.am	Thu Nov 28 00:46:47 2013
@@ -8,7 +8,7 @@
 SET_DEBUG_OPTS = @SET_DEBUG_OPTS@
 SET_INCLUDES = -I$(top_srcdir) -Ifile -Ilibbluray/bdnav
 
-AM_CFLAGS = -std=c99 $(SET_FEATURES) $(SET_WARNINGS) $(SET_OPTIMIZATIONS) \
+AM_CFLAGS = -std=gnu99 $(SET_FEATURES) $(SET_WARNINGS) $(SET_OPTIMIZATIONS) \
         $(SET_DEBUG_OPTS) $(SET_INCLUDES) $(LIBXML2_CFLAGS)
 CFLAGS=$(shell test -z "$$CFLAGS" || echo "$$CFLAGS")
 
--- libbluray-0.2.3/src/Makefile.in.bak	Fri Aug 17 07:31:46 2012
+++ libbluray-0.2.3/src/Makefile.in	Thu Nov 28 00:46:47 2013
@@ -353,7 +353,7 @@
 SUBDIRS = . examples
 SET_FEATURES = -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_REENTRANT
 SET_INCLUDES = -I$(top_srcdir) -Ifile -Ilibbluray/bdnav
-AM_CFLAGS = -std=c99 $(SET_FEATURES) $(SET_WARNINGS) \
+AM_CFLAGS = -std=gnu99 $(SET_FEATURES) $(SET_WARNINGS) \
 	$(SET_OPTIMIZATIONS) $(SET_DEBUG_OPTS) $(SET_INCLUDES) \
 	$(LIBXML2_CFLAGS) $(am__append_2)
 SOURCES_bdj = libbluray/bdj/bdj.h \
--- libbluray-0.2.3/configure.ac.bak	Fri Aug 17 07:27:32 2012
+++ libbluray-0.2.3/configure.ac	Thu Nov 28 01:18:14 2013
@@ -163,13 +163,13 @@
 # function testing for supported compiler options
 check_cc_options()
 {
-  local tmpfile=$(mktemp XXXXXXXXXX)
-  local tmpcfile="$tmpfile.c"
+  tmpfile=$(mktemp XXXXXXXXXX)
+  tmpcfile="$tmpfile.c"
   mv $tmpfile $tmpcfile
-  local retval=1
+  retval=1
   printf "Checking if compiler supports $@... "
   if $CC -c $@ $tmpcfile -o /dev/null >/dev/null 2>&1; then
-    local retval=0
+    retval=0
   fi
   rm $tmpcfile
   if [[ $retval = "0" ]]; then
