diff --git a/src/libbluray/disc/disc.c b/src/libbluray/disc/disc.c
index 6d555a1..dd8e5a5 100644
--- a/src/libbluray/disc/disc.c
+++ b/src/libbluray/disc/disc.c
@@ -283,7 +283,7 @@ static BD_DISC *_disc_init()
 static void _set_paths(BD_DISC *p, const char *device_path)
 {
     if (device_path) {
-        char *disc_root = mount_get_mountpoint(device_path);
+        char *disc_root = str_dup(device_path);
 
         /* make sure path ends to slash */
         if (!disc_root || (disc_root[0] && disc_root[strlen(disc_root) - 1] == DIR_SEP_CHAR)) {