Index: mp4v2/include/mp4v2/platform.h
===================================================================
--- mp4v2/include/mp4v2/platform.h	(revision 272)
+++ mp4v2/include/mp4v2/platform.h	(working copy)
@@ -24,8 +24,10 @@
 #if defined( _WIN32 ) || defined( __MINGW32__ )
 #   if defined( _WINDLL ) || defined( DLL_EXPORT )
 #       define MP4V2_EXPORT __declspec(dllexport)
+#   elif defined( _DLL ) || defined( DLL_IMPORT ) 
+#       define MP4V2_EXPORT __declspec(dllimport)
 #   else
-#       define MP4V2_EXPORT __declspec(dllimport)
+#       define MP4V2_EXPORT
 #   endif
 #else
 #   define MP4V2_EXPORT __attribute__((visibility("default")))
