diff -Naur libmp4v2.orig/configure.ac libmp4v2/configure.ac
--- libmp4v2.orig/configure.ac	2009-01-10 12:44:40.000000000 -0500
+++ libmp4v2/configure.ac	2009-03-13 18:18:38.000000000 -0400
@@ -103,11 +103,11 @@
         X_CXX_ARCH="$X_CXX_ARCH -mwin32"
     fi
     ;;
-*-*-mingw32)
+*-*-mingw*)
     # do not support shared
     enable_shared=no
     X_PLATFORM=win32
-    X_MINGW_LIBS="$X_MINGW_LIBS -lmsvcr80"
+    X_MINGW_LIBS="$X_MINGW_LIBS"
     X_CXX_W="$X_CXX_W -Wno-format"
     if test "$enable_mingw_threads" = "yes"; then
         X_CXX_ARCH="$X_CXX_ARCH -mthreads"
@@ -197,7 +197,7 @@
     case ${host} in
         *-*-cygwin)
             ;;
-        *-*-mingw32)
+        *-*-mingw*)
             ;;
         *)
             AC_LANG(C++)
