diff -Naur libdvdread.orig/src/dvd_udf.c libdvdread/src/dvd_udf.c
--- libdvdread.orig/src/dvd_udf.c	2009-01-08 14:57:10.000000000 -0800
+++ libdvdread/src/dvd_udf.c	2009-04-23 13:36:08.000000000 -0700
@@ -928,7 +928,7 @@
   if(GetUDFCache(device, PVDCache, 0, pvd))
     return 1;
 
-  if(!UDFGetDescriptor( device, 1, pvd_buf, sizeof(pvd_buf)))
+  if(!UDFGetDescriptor( device, 1, pvd_buf, DVD_VIDEO_LB_LEN))
     return 0;
 
   memcpy(pvd->VolumeIdentifier, &pvd_buf[24], 32);
