diff --git a/src/dvd_input.c b/src/dvd_input.c
index 0c44387..476c9c8 100644
--- a/src/dvd_input.c
+++ b/src/dvd_input.c
@@ -319,6 +319,7 @@ int dvdinput_setup(void *priv, dvd_logger_cb *logcb)
 
 #ifdef __APPLE__
   #define CSS_LIB "libdvdcss.2.dylib"
+  #define CSS_USR_LOCAL_LIB "/usr/local/lib/libdvdcss.2.dylib"
 #elif defined(_WIN32)
   #define CSS_LIB "libdvdcss-2.dll"
 #elif defined(__OS2__)
@@ -328,6 +329,13 @@ int dvdinput_setup(void *priv, dvd_logger_cb *logcb)
 #endif
   dvdcss_library = dlopen(CSS_LIB, RTLD_LAZY);
 
+#ifdef __APPLE__
+  // Use the absolute path, required by the hardened runtime
+  if(dvdcss_library == NULL) {
+    dvdcss_library = dlopen(CSS_USR_LOCAL_LIB, RTLD_LAZY);
+  }
+#endif
+
   if(dvdcss_library != NULL) {
 #if defined(__OpenBSD__) && !defined(__ELF__) || defined(__OS2__)
 #define U_S "_"
