diff --git a/src/dvdread/ifo_types.h b/src/dvdread/ifo_types.h
index f679d29..699853f 100644
--- a/src/dvdread/ifo_types.h
+++ b/src/dvdread/ifo_types.h
@@ -30,7 +30,7 @@
 #undef PRAGMA_PACK_BEGIN
 #undef PRAGMA_PACK_END
 
-#if defined(__GNUC__)
+#if defined(__GNUC__) && !defined(__clang__)
 #if __GNUC__ > 2 || (__GNUC__ == 2 && __GNUC_MINOR__ >= 95)
 #if __GNUC__ > 3 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 4)
 #define ATTRIBUTE_PACKED __attribute__ ((packed,gcc_struct))