diff --git a/src/dvd_input.c b/src/dvd_input.c
--- a/src/dvd_input.c	2025-09-25 21:59:39.000000000 +0200
+++ b/src/dvd_input.c	2025-09-29 19:21:29.888738600 +0200
@@ -479,6 +479,7 @@
 
 #ifdef __APPLE__
   #define CSS_LIB "libdvdcss.2.dylib"
+  #define CSS_USR_LOCAL_LIB "/usr/local/lib/libdvdcss.2.dylib"
 #elif defined(_WIN32)
   #define CSS_LIB "libdvdcss-2.dll"
 #elif defined(__OS2__)
@@ -515,6 +516,13 @@
   dvdcss_library = dlopen(CSS_LIB, RTLD_LAZY);
 #endif
 
+#ifdef __APPLE__
+  // Use the absolute path, required by the hardened runtime
+  if(dvdcss_library == NULL) {
+    dvdcss_library = dlopen(CSS_USR_LOCAL_LIB, RTLD_LAZY);
+  }
+#endif
+
   if(dvdcss_library != NULL) {
 #ifdef __OS2__
 #define U_S "_"
