Index: configure.ac
===================================================================
--- libdvdread/configure.ac	(revision 1168)
+++ libdvdread/configure.ac	(working copy)
@@ -158,6 +158,16 @@
     ;;
 esac
 
+AC_ARG_ENABLE([local-dlfcn],
+  [AS_HELP_STRING([--enable-local-dlfcn],
+  [use local dlfcn for mingw (default is auto)])],
+  [use_local_dlfcn=$enableval],
+  [use_local_dlfcn=no])
+
+if [[ $use_local_dlfcn = "yes" ]]; then
+	AC_DEFINE([USING_LOCAL_DLFCN], [1], ["Define to 1 to use local dlfcn"])
+fi
+
 dnl ---------------------------------------------
 dnl cflags
 dnl ---------------------------------------------
Index: src/dvd_input.c
===================================================================
--- libdvdread/src/dvd_input.c	(revision 1168)
+++ libdvdread/src/dvd_input.c	(working copy)
@@ -50,7 +50,7 @@
 #else
 
 /* dlopening libdvdcss */
-#ifdef HAVE_DLFCN_H
+#if defined(HAVE_DLFCN_H) && !defined(USING_LOCAL_DLFCN)
 #include <dlfcn.h>
 #else
 /* Only needed on MINGW at the moment */
