diff -Naur libdvdread.orig/src/dvd_input.c libdvdread/src/dvd_input.c
--- libdvdread.orig/src/dvd_input.c	2009-01-08 14:57:10.000000000 -0800
+++ libdvdread/src/dvd_input.c	2009-04-24 09:02:34.000000000 -0700
@@ -285,7 +285,7 @@
   /* dlopening libdvdcss */
 
 #ifdef __APPLE__
-  #define CSS_LIB "libdvdcss.2.dylib"
+  #define CSS_LIB "/Applications/VLC.app/Contents/MacOS/lib/libdvdcss.2.dylib"
 #elif defined(WIN32)
   #define CSS_LIB "libdvdcss.dll"
 #elif defined(__OS2__)
