diff -Naur libdvdread.orig/src/dvd_input.h libdvdread/src/dvd_input.h
--- libdvdread.orig/src/dvd_input.h	2008-10-03 13:11:30.000000000 -0700
+++ libdvdread/src/dvd_input.h	2009-04-23 13:47:04.000000000 -0700
@@ -29,6 +29,24 @@
 
 #define DVDINPUT_READ_DECRYPT    (1 << 0)
 
+#if defined( __MINGW32__ )
+#   undef  lseek
+#   define lseek  _lseeki64
+#   undef  fseeko
+#   define fseeko fseeko64
+#   undef  ftello
+#   define ftello ftello64
+#   define flockfile(...)
+#   define funlockfile(...)
+#   define getc_unlocked getc
+#   undef  off_t
+#   define off_t off64_t
+#   undef  stat
+#   define stat  _stati64
+#   define fstat _fstati64
+#   define wstat _wstati64
+#endif
+
 typedef struct dvd_input_s *dvd_input_t;
 
 /**
