--- libdvdread.orig/dvdread/dvd_input.h	2005-09-06 12:23:00.000000000 -0700
+++ libdvdread/dvdread/dvd_input.h	2009-03-25 15:01:39.000000000 -0700
@@ -28,6 +28,24 @@
 
 #define DVDINPUT_READ_DECRYPT    (1 << 0)
 
+#if defined( __MINGW32__ )
+#   undef  lseek
+#   define lseek  _lseeki64
+#   undef  fseeko
+#   define fseeko fseeko64
+#   undef  ftello
+#   define ftello ftello64
+#   define flockfile(...)
+#   define funlockfile(...)
+#   define getc_unlocked getc
+#   undef  off_t
+#   define off_t off64_t
+#   undef  stat
+#   define stat  _stati64
+#   define fstat _fstati64
+#   define wstat _wstati64
+#endif
+
 typedef struct dvd_input_s *dvd_input_t;
 
 /**

