diff --git a/source/CMakeLists.txt b/source/CMakeLists.txt
index 60031df63..7120f9414 100755
--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -260,7 +260,13 @@ if(GCC)
         endif()
     endif()
 	if(ARM64 OR CROSS_COMPILE_ARM64)
-	    set(ARM_ARGS -fPIC -flax-vector-conversions)
+        if (MINGW)
+            set(ARM_ARGS -flax-vector-conversions)
+        elseif(APPLE)
+            set(ARM_ARGS -fPIC -flax-vector-conversions -arch ${CMAKE_OSX_ARCHITECTURES})
+        else()
+            set(ARM_ARGS -fPIC -flax-vector-conversions)
+        endif()
         find_package(Neon)
         if(CPU_HAS_NEON)
             add_definitions(-DHAVE_NEON)
