diff -r 32d72a0285e0 source/common/param.cpp
--- a/source/common/param.cpp	Tue Feb 10 21:38:22 2015 -0600
+++ b/source/common/param.cpp	Thu Feb 12 09:12:01 2015 -0700
@@ -1151,6 +1151,7 @@
 
 int x265_set_globals(x265_param *param)
 {
+#if 0
     static int once /* = 0 */;
 
     if (ATOMIC_INC(&once) > 1)
@@ -1162,6 +1163,7 @@
         }
     }
     else
+#endif
     {
         uint32_t maxLog2CUSize = (uint32_t)g_log2Size[param->maxCUSize];
 
