diff -Ncr x265-old/source/common/common.h x265-new/source/common/common.h
*** x265-old/source/common/common.h	Wed Jan 15 09:08:21 2014
--- x265-new/source/common/common.h	Thu Jan 23 15:16:39 2014
***************
*** 27,32 ****
--- 27,42 ----
  #include <cstdlib>
  #include "x265.h"
  
+ const x265_cli_csp x265_cli_csps[] =
+ {
+     { 1, { 0, 0, 0 }, { 0, 0, 0 } }, /* i400 */
+     { 3, { 0, 1, 1 }, { 0, 1, 1 } }, /* i420 */
+     { 3, { 0, 1, 1 }, { 0, 0, 0 } }, /* i422 */
+     { 3, { 0, 0, 0 }, { 0, 0, 0 } }, /* i444 */
+     { 2, { 0, 0 },    { 0, 1 } },    /* nv12 */
+     { 2, { 0, 0 },    { 0, 0 } },    /* nv16 */
+ };
+ 
  #define X265_MIN(a, b) ((a) < (b) ? (a) : (b))
  #define X265_MAX(a, b) ((a) > (b) ? (a) : (b))
  #define COPY1_IF_LT(x, y) if ((y) < (x)) (x) = (y);
diff -Ncr x265-old/source/x265.h x265-new/source/x265.h
*** x265-old/source/x265.h	Wed Jan 15 09:08:21 2014
--- x265-new/source/x265.h	Thu Jan 23 15:16:23 2014
***************
*** 195,210 ****
      int height[3];
  } x265_cli_csp;
  
- const x265_cli_csp x265_cli_csps[] =
- {
-     { 1, { 0, 0, 0 }, { 0, 0, 0 } }, /* i400 */
-     { 3, { 0, 1, 1 }, { 0, 1, 1 } }, /* i420 */
-     { 3, { 0, 1, 1 }, { 0, 0, 0 } }, /* i422 */
-     { 3, { 0, 0, 0 }, { 0, 0, 0 } }, /* i444 */
-     { 2, { 0, 0 },    { 0, 1 } },    /* nv12 */
-     { 2, { 0, 0 },    { 0, 0 } },    /* nv16 */
- };
- 
  /* rate tolerance method */
  typedef enum
  {
--- 195,200 ----
