From ae2c6b8c2b665c3be30649e8ce68d39be7cbbcd3 Mon Sep 17 00:00:00 2001
From: Max Dmitrichenko <max.dmitrichenko@intel.com>
Date: Thu, 22 Aug 2024 07:40:48 +0200
Subject: [PATCH 1/6] Do not set thread priority on Windows

---
 source/common/threadpool.cpp | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/source/common/threadpool.cpp b/source/common/threadpool.cpp
index ec96f18b1..615b04b6f 100644
--- a/source/common/threadpool.cpp
+++ b/source/common/threadpool.cpp
@@ -115,12 +115,6 @@ void WorkerThread::threadMain()
 {
     THREAD_NAME("Worker", m_id);
 
-#if _WIN32
-    SetThreadPriority(GetCurrentThread(), THREAD_PRIORITY_BELOW_NORMAL);
-#else
-    __attribute__((unused)) int val = nice(10);
-#endif
-
     m_pool.setCurrentThreadAffinity();
 
     sleepbitmap_t idBit = (sleepbitmap_t)1 << m_id;
-- 
2.39.5 (Apple Git-154)

