From 99ff0f140e700f0cce5d94996cde4b1b9a325488 Mon Sep 17 00:00:00 2001
From: Max Dmitrichenko <max.dmitrichenko@intel.com>
Date: Thu, 22 Aug 2024 07:40:48 +0200
Subject: [PATCH 1/7] Do not set thread priority on Windows

---
 source/common/threadpool.cpp | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/source/common/threadpool.cpp b/source/common/threadpool.cpp
index 6bde73c1d..140445119 100644
--- a/source/common/threadpool.cpp
+++ b/source/common/threadpool.cpp
@@ -125,12 +125,6 @@ void WorkerThread::threadMain()
 {
     THREAD_NAME("Worker", m_id);
 
-#if _WIN32
-    SetThreadPriority(GetCurrentThread(), THREAD_PRIORITY_BELOW_NORMAL);
-#else
-    __attribute__((unused)) int val = nice(10);
-#endif
-
     m_pool.setCurrentThreadAffinity();
 
     sleepbitmap_t idBit = (sleepbitmap_t)1 << m_id;
-- 
2.50.1 (Apple Git-155)

