From 4a56a4255cbbc008c1973eb74ab26bfbea204103 Mon Sep 17 00:00:00 2001
From: Damiano Galassi <damiog@gmail.com>
Date: Fri, 22 Nov 2024 11:05:34 +0100
Subject: [PATCH] Simplify macOS cross-compilation

---
 source/CMakeLists.txt | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/source/CMakeLists.txt b/source/CMakeLists.txt
index 5c6dda9e8..a31cde199 100755
--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -736,6 +736,9 @@ if((MSVC_IDE OR XCODE OR GCC) AND ENABLE_ASSEMBLY)
    elseif(ARM64 OR CROSS_COMPILE_ARM64)
     # compile ARM64 arch asm files here
         enable_language(ASM)
+        if(APPLE)
+            set(ARM_ARGS ${ARM_ARGS} -arch ${CMAKE_OSX_ARCHITECTURES})
+        endif()
         foreach(ASM ${ARM_ASMS})
             set(ASM_SRC ${CMAKE_CURRENT_SOURCE_DIR}/common/aarch64/${ASM})
             list(APPEND ASM_SRCS ${ASM_SRC})
-- 
2.39.5 (Apple Git-154)

