From 0d8bab0c3c5ec3b5676821f95b0e279a9713d4bf Mon Sep 17 00:00:00 2001
From: Damiano Galassi <damiog@gmail.com>
Date: Fri, 13 Sep 2024 16:10:16 +0200
Subject: [PATCH 6/6] Simplify macOS cross-compilation

---
 source/CMakeLists.txt | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/source/CMakeLists.txt b/source/CMakeLists.txt
index 37b83f959..f14fc0bd1 100755
--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -334,7 +334,11 @@ if(GCC)
             set(ARM_ARGS -O3 -march=armv9-a+i8mm+sve2)
             add_definitions(-DHAVE_SVE2=1)
         endif()
-        set(ARM_ARGS ${ARM_ARGS} -fPIC)
+        if(APPLE)
+            set(ARM_ARGS ${ARM_ARGS} -fPIC -arch ${CMAKE_OSX_ARCHITECTURES})
+        else()
+            set(ARM_ARGS ${ARM_ARGS} -fPIC)
+        endif()
         # Do not allow implicit vector type conversions in Clang builds (this
         # is already the default in GCC builds).
         check_cxx_compiler_flag(-flax-vector-conversions=none CC_HAS_FLAX_VEC_CONV_NONE)
-- 
2.39.3 (Apple Git-146)

