diff --git a/source/common/aarch64/pixel-util.S b/source/common/aarch64/pixel-util.S
index fba9a90d5..acccd28fe 100644
--- a/source/common/aarch64/pixel-util.S
+++ b/source/common/aarch64/pixel-util.S
@@ -2407,7 +2407,7 @@ function PFX(costCoeffNxN_neon)
     // x5 - scanFlagMask
     // x6 - baseCtx
     mov             x0, #0
-    movrel          x1, x265_entropyStateBits
+    movrel          x1, _x265_entropyStateBits
     mov             x4, #0
     mov             x11, #0
     movi            v31.16b, #0
