--- a/x264.h.orig	2018-11-20 06:39:30.000000000 -0500
+++ b/x264.h	2018-11-20 06:39:07.000000000 -0500
@@ -45,7 +45,24 @@
 
 #include "x264_config.h"
 
-#define X264_BUILD 157
+/*
+ * Define the full version explicitly so that it survives a git --archive.
+ *
+ * Patch will break every time X264_BUILD changes, so that we don't forget to
+ * update it ;-)
+ */
+#ifdef  X264_VERSION
+#undef  X264_VERSION
+#endif
+#ifdef  X264_BUILD
+#undef  X264_BUILD
+#endif
+#ifdef  X264_POINTVER
+#undef  X264_POINTVER
+#endif
+#define X264_BUILD    157
+#define X264_VERSION " r2935 545de2f"
+#define X264_POINTVER "0.157.2935 545de2f"
 
 /* Application developers planning to link against a shared library version of
  * libx264 from a Microsoft Visual Studio or similar development environment
