diff --git a/x264.h b/x264.h
index 2b59b92..6211a64 100644
--- a/x264.h
+++ b/x264.h
@@ -45,7 +45,17 @@ extern "C" {
 
 #include "x264_config.h"
 
-#define X264_BUILD 148
+/*
+ * Define the full version explicitly so that it survives a git --archive.
+ *
+ * Patch will break every time X264_BUILD changes, so that we don't forget to
+ * update it ;-)
+ */
+#ifdef  X264_VERSION
+#undef  X264_VERSION
+#endif
+#define X264_BUILD       148
+#define X264_VERSION " r2705 3f5ed56"
 
 /* Application developers planning to link against a shared library version of
  * libx264 from a Microsoft Visual Studio or similar development environment
