diff --git a/x264.h b/x264.h
index d685efbd..7f270c4b 100644
--- a/x264.h
+++ b/x264.h
@@ -45,7 +45,24 @@ extern "C" {
 
 #include "x264_config.h"
 
-#define X264_BUILD 164
+/*
+ * Define the full version explicitly so that it survives a git --archive.
+ *
+ * Patch will break every time X264_BUILD changes, so that we don't forget to
+ * update it ;-)
+ */
+#ifdef  X264_VERSION
+#undef  X264_VERSION
+#endif
+#ifdef  X264_BUILD
+#undef  X264_BUILD
+#endif
+#ifdef  X264_POINTVER
+#undef  X264_POINTVER
+#endif
+#define X264_BUILD    164
+#define X264_VERSION " r3100 ed0f7a6"
+#define X264_POINTVER "0.164.3100 ed0f7a6"
 
 #ifdef _WIN32
 #   define X264_DLL_IMPORT __declspec(dllimport)
