diff --git a/x264.h b/x264.h
index 2b59b92..86b7198 100644
--- a/x264.h
+++ b/x264.h
@@ -45,7 +45,24 @@ extern "C" {
 
 #include "x264_config.h"
 
-#define X264_BUILD 155
+/*
+ * Define the full version explicitly so that it survives a git --archive.
+ *
+ * Patch will break every time X264_BUILD changes, so that we don't forget to
+ * update it ;-)
+ */
+#ifdef  X264_VERSION
+#undef  X264_VERSION
+#endif
+#ifdef  X264_BUILD
+#undef  X264_BUILD
+#endif
+#ifdef  X264_POINTVER
+#undef  X264_POINTVER
+#endif
+#define X264_BUILD    155
+#define X264_VERSION " r2901 7d0ff22"
+#define X264_POINTVER "0.155.2901 7d0ff22"
 
 /* Application developers planning to link against a shared library version of
  * libx264 from a Microsoft Visual Studio or similar development environment
