diff -ru x264/Makefile x264.solaris/Makefile
--- x264/Makefile	Mon Jun  9 01:36:08 2008
+++ x264.solaris/Makefile	Thu Jun 19 11:54:53 2008
@@ -150,8 +150,10 @@
 	$(MAKE) -C gtk distclean
 
 install: x264$(EXE) $(SONAME)
-	install -d $(DESTDIR)$(bindir) $(DESTDIR)$(includedir)
-	install -d $(DESTDIR)$(libdir) $(DESTDIR)$(libdir)/pkgconfig
+	install -d $(DESTDIR)$(bindir) 
+	install -d $(DESTDIR)$(includedir)
+	install -d $(DESTDIR)$(libdir) 
+	install -d $(DESTDIR)$(libdir)/pkgconfig
 	install -m 644 x264.h $(DESTDIR)$(includedir)
 	install -m 644 libx264.a $(DESTDIR)$(libdir)
 	install -m 644 x264.pc $(DESTDIR)$(libdir)/pkgconfig
diff -ru x264/configure x264.solaris/configure
--- x264/configure	Tue Aug 12 03:26:58 2008
+++ x264.modified/configure	Tue Aug 19 16:09:32 2008
@@ -329,8 +327,10 @@
 
 echo "unsigned int endian = 'B' << 24 | 'I' << 16 | 'G' << 8 | 'E';" > conftest.c
 $CC $CFLAGS conftest.c -c -o conftest.o 2>$DEVNULL || die "endian test failed"
-grep -q BIGE conftest.o && CFLAGS="$CFLAGS -DWORDS_BIGENDIAN"
+/usr/bin/tr -cd "BIGE" < conftest.o | /usr/xpg4/bin/grep -q BIGE && CFLAGS="$CFLAGS -DWORDS_BIGENDIAN"
 
+echo $CFLAGS
+
 # autodetect options that weren't forced nor disabled
 
 libpthread=""
