diff --git a/src/mfx_library_iterator_linux.cpp b/src/mfx_library_iterator_linux.cpp
index 46ad511..e2269f3 100644
--- a/src/mfx_library_iterator_linux.cpp
+++ b/src/mfx_library_iterator_linux.cpp
@@ -332,8 +332,7 @@ mfxStatus MFXLibraryIterator::Init(eMfxImplType implType, mfxIMPL impl, const mf
     // set the required library's implementation type
     m_implType = implType;
 
-    snprintf(m_path, sizeof(m_path)/sizeof(m_path[0]),
-             "%s/%s/%04x/%04x", mfx_storage_opt, mfx_folder, m_vendorID, m_deviceID);
+    snprintf(m_path, sizeof(m_path)/sizeof(m_path[0]),"%s/%s", mfx_storage_opt, mfx_folder);
 
     m_libs_num = mfx_list_libraries(m_path, (MFX_LIB_HARDWARE == implType), &m_libs);
 
