--- mpeg2dec/configure.bak	Fri Jul 18 10:41:54 2008
+++ mpeg2dec/configure	Wed Nov 27 23:04:30 2013
@@ -4614,7 +4614,7 @@
 elif test x"$GCC" = x"yes"; then
 
 
-            TRY_CFLAGS="$OPT_CFLAGS -Wall -Wpointer-arith -Wcast-align -Wcast-qual -Wstrict-prototypes -Wshadow -Waggregate-return -Wmissing-prototypes -Wnested-externs -Wsign-compare"
+            TRY_CFLAGS="$OPT_CFLAGS -Wall -Wpointer-arith -Wcast-align -Wcast-qual  -Wshadow -Waggregate-return -Wmissing-prototypes -Wnested-externs -Wsign-compare"
     { echo "$as_me:$LINENO: checking if $CC supports $TRY_CFLAGS $CFLAGS flags" >&5
 echo $ECHO_N "checking if $CC supports $TRY_CFLAGS $CFLAGS flags... $ECHO_C" >&6; }
     SAVE_CFLAGS="$CFLAGS"
@@ -4859,11 +4859,11 @@
 _ACEOF
 
 	case "$host" in
-	i386-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=i386";;
-	i486-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=i486";;
-	i586-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=pentium";;
-	i686-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=pentiumpro";;
-	k6-*)   TRY_CFLAGS="$OPT_CFLAGS -mcpu=k6";;
+	i386-*) TRY_CFLAGS="$OPT_CFLAGS -mtune=i386";;
+	i486-*) TRY_CFLAGS="$OPT_CFLAGS -mtune=i486";;
+	i586-*) TRY_CFLAGS="$OPT_CFLAGS -mtune=pentium";;
+	i686-*) TRY_CFLAGS="$OPT_CFLAGS -mtune=pentiumpro";;
+	k6-*)   TRY_CFLAGS="$OPT_CFLAGS -mtune=k6";;
 	esac
 	{ echo "$as_me:$LINENO: checking if $CC supports $TRY_CFLAGS $CFLAGS flags" >&5
 echo $ECHO_N "checking if $CC supports $TRY_CFLAGS $CFLAGS flags... $ECHO_C" >&6; }
@@ -5084,7 +5084,7 @@
 #define ARCH_SPARC
 _ACEOF
 
-	TRY_CFLAGS="$OPT_CFLAGS -mcpu=ultrasparc -mvis"
+	TRY_CFLAGS="$OPT_CFLAGS -mtune=ultrasparc -mvis"
 	{ echo "$as_me:$LINENO: checking if $CC supports $TRY_CFLAGS $CFLAGS flags" >&5
 echo $ECHO_N "checking if $CC supports $TRY_CFLAGS $CFLAGS flags... $ECHO_C" >&6; }
     SAVE_CFLAGS="$CFLAGS"
