From 39a7021f96bc49d03fb5b84d1daf31074878f98d Mon Sep 17 00:00:00 2001
From: Luca Barbato <lu_zero@gentoo.org>
Date: Mon, 26 Oct 2020 14:20:00 +0100
Subject: [PATCH] Add an explicit feature to vendor openssl

---
 Cargo.toml | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Cargo.toml b/Cargo.toml
index 454c31b..742bf7b 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -33,3 +33,7 @@ toml = "0.5"
 serde = "1.0"
 serde_derive = "1.0"
 anyhow = "1.0"
+
+
+[features]
+vendored-openssl = ["cargo/vendored-openssl"]
